mindmeld/components/question_answerer.py

```diff
@@ -449,805 +591,2172 @@ def _generate_mapping_data(embedder_model, embedding_fields):
+                for key, filter_text in kwargs.items():
+                    if key in self._filter_queries:
+                        msg = f"Found a duplicate filter clause against '{key}' field name. " \
+                              "Utilizing only latest input."
+                        logger.warning(msg)
+                    self._filter_queries.update({key: {"filter_text": filter_text}})
 
-                es_query["query"]["function_score"]["query"]["bool"][
-                    "filter"
-                ] = es_filter_clauses
+            return self
 
-        # add scoring function for custom sort criteria
-        for clause in self._clauses["sort"]:
-            sort_function = clause.build_query()
-            es_query["query"]["function_score"]["functions"].append(sort_function)
+        def sort(self, field, sort_type=None, location=None):
 
-        logger.debug("ES query syntax: %s.", es_query)
+            if field in self._sort_queries:
+                msg = f"Found a duplicate sort clause against '{field}' field name. " \
+                      "Utilizing only latest input."
+                logger.warning(msg)
+            self._sort_queries.update(
+                {field: {"sort_type": sort_type, "location": location}})
 
-        return es_query
+            return self
 
-    def execute(self, size=10):
-        """Executes the knowledge base search with provided criteria and returns matching documents.
+        def execute(self, size=10):
 
-        Args:
-            size (int): The maximum number of records to fetch, default to 10.
+            try:
+                index_resources = NonElasticsearchQuestionAnswerer.ALL_INDICES.get(self.index_name)
+                # obtain all doc ids in the order they were present in the KB
+                index_all_ids = (
+                    NonElasticsearchQuestionAnswerer.ALL_INDICES.get_all_ids(self.index_name)
+                )
+            except KeyError:
+                msg = f"The index '{self.index_name}' looks unavailable. " \
+                      f"Consider running '.load_kb(...)' to create indices " \
+                      f"before creating search/filter/sort queries. "
+                logger.error(msg)
+                return []
+
+            # get results (aka. curated_docs) for 'query' clauses, in decreasing order of similarity
+            n_scores = 0
+            scores = {}
+            for field_name, kwargs in self._search_queries.items():
+                query_type = kwargs["query_type"]
+                value = kwargs["value"]
+                field_resource = index_resources.get(field_name)
+                if not field_resource:
+                    msg = f"The field '{field_name}' is not available in index '{self.index_name}'."
+                    logger.error(msg)
+                    raise ValueError("Invalid knowledge base field '{}'".format(field_name))
+                this_field_scores = field_resource.do_search(query_type, value)
```

Comment:
not sure if i'm missing something but we probably should apply filters first (if specified) so that we only need to compute scores on a narrower set?

Reply:
Thanks for pointing it out. Makes sense to do filtering first. Will make this change in code.